plugins/outputs/postgresql/table_manager.go

```diff
@@ -0,0 +1,388 @@
+	if err != nil {
+		return fmt.Errorf("refreshing table: %w", err)
+	}
+	tm.refreshTableStructureResponse(tableName, rows)
+	*/
+
+	// Lock to prevent concurrency issues in postgres (pg_type_typname_nsp_index unique constraint; SQLSTATE 23505)
+	tm.schemaMutex.Lock()
+	defer tm.schemaMutex.Unlock()
+
+	tx, err := db.Begin(ctx)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback(ctx)
+
+	for _, tmpl := range tmpls {
+		sql, err := tmpl.Render(tmplTable, newColumns, metricsTmplTable, tagsTmplTable)
+		if err != nil {
+			return err
+		}
+		if _, err := tx.Exec(ctx, string(sql)); err != nil {
+			return fmt.Errorf("executing `%s`: %w", sql, err)
+		}
+	}
+
+	// We need to be able to determine the role of the column when reading the structure back (because of the templates).
+	// For some columns we can determine this by the column name (time, tag_id, etc). However tags and fields can have any
+	// name, and look the same. So we add a comment to tag columns, and through process of elimination what remains are
+	// field columns.
+	for _, col := range newColumns {
+		if col.Role != utils.TagColType {
+			continue
+		}
+		if _, err := tx.Exec(ctx, "COMMENT ON COLUMN "+tmplTable.String()+"."+template.QuoteIdentifier(col.Name)+" IS 'tag'"); err != nil {
+			return fmt.Errorf("setting column role comment: %s", err)
+		}
+	}
+
+	if err := tx.Commit(ctx); err != nil {
+		return err
+	}
+
+	return tm.refreshTableStructure(ctx, db, tbl)
+}
+
+func colMapToSlice(colMap map[string]utils.Column) []utils.Column {
+	if colMap == nil {
+		return nil
+	}
+	cols := make([]utils.Column, 0, len(colMap))
+	for _, col := range colMap {
+		cols = append(cols, col)
+	}
+	return cols
+}
+
+// MatchSource scans through the metrics, determining what columns are needed for inserting, and ensuring the DB schema matches.
+//
+// If the schema does not match, and schema updates are disabled:
+// If a field missing from the DB, the field is omitted.
+// If a tag is missing from the DB, the metric is dropped.
+func (tm *TableManager) MatchSource(ctx context.Context, db dbh, rowSource *TableSource) error {
+	metricTable := tm.table(rowSource.Name())
+	var tagTable *tableState
+	if tm.TagsAsForeignKeys {
+		tagTable = tm.table(metricTable.name + tm.TagTableSuffix)
+
+		missingCols, err := tm.EnsureStructure(
+			ctx,
+			db,
+			tagTable,
+			rowSource.TagTableColumns(),
+			tm.TagTableCreateTemplates,
+			tm.TagTableAddColumnTemplates,
+			metricTable,
+			tagTable,
+		)
+		if err != nil {
+			return err
+		}
+
+		if len(missingCols) > 0 {
+			colDefs := make([]string, len(missingCols))
+			for i, col := range missingCols {
+				if err := rowSource.DropColumn(col); err != nil {
+					return fmt.Errorf("metric/table mismatch: Unable to omit field/column from \"%s\": %w", tagTable.name, err)
+				}
+				colDefs[i] = col.Name + " " + string(col.Type)
+			}
+			tm.Logger.Errorf("table '%s' is missing tag columns (dropping metrics): %s", tagTable.name, strings.Join(colDefs, ", "))
```

Comment:
Does it actually make sense to operate the plugin in that state or should we return an error here?

Reply:
Yes. If schema updates are disabled, then this situation will occur. When that happens, if we start returning errors, valid metrics won't make it through.